src/transformStateWithClones.js

```diff
@@ -7,7 +7,36 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let currentState = state;
+  const actionHistory = [];
+
+  for (const action of actions) {
+    const newState = Object.assign({}, currentState);
+
+    switch (action.type) {
+      case 'addProperties':
+        Object.assign(newState, action.extraData);
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete newState[key];
+        }
+        break;
+
+      case 'clear':
+        for (const key in newState) {
+          delete newState[key];
+        }
```

Comment:
Minor, but you can probably skip deleting all keys by just putting empty object into the history

```suggestion
            actionHistory.push({});
            currentState = {}
            
            // go to next loop iteration
```